### PR TITLE
Reduces code used for e2e tests

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This removes a couple custom types we used in e2e tests. We don't need to work so much since there are built-in functions that allow us to accomplish the same. This also cleans up where we accidentally left run archives around after running e2e tests.

### CODE REVIEW

File: e2e/getenvoy_run_test.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 Tetrate
+// Copyright 2021 Tetrate
```
Comment (codefromthecrypt) [inline e2e/getenvoy_run_test.go:1]: because there's basically nothing common in this file anymore

File: e2e/getenvoy_test.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 Tetrate
+// Copyright 2021 Tetrate
```
Comment (codefromthecrypt) [inline e2e/getenvoy_test.go:1]: this was a typo as there was never a version command before this year :D

File: e2e/main_test.go [hunk 3]
```diff
@@ -124,3 +131,29 @@ func readGetEnvoyPath() (string, error) {
...
 }
+
+// getEnvoy's main function is providing a nice String
+type getEnvoy struct {
+	cmd              *exec.Cmd
+	adminAddressPath string
+}
+
+func newGetEnvoy(args ...string) (*getEnvoy, func()) {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
```
Comment (codefromthecrypt) [inline e2e/main_test.go:142]: this is cool

File: e2e/getenvoy_run_test.go [hunk 4]
```diff
@@ -30,113 +33,139 @@ import (
...
+	err = c.cmd.Start()
+	require.NoError(t, err)
+
+	log.Printf(`waiting for getenvoy stdout to match %q after running [%v]`, adminAddressPathPattern, c)
+	go func() {
+		for outLines.Scan() {
+			l := outLines.Text()
+			fmt.Fprintln(os.Stdout, l) // echo to the person watching the test
+
+			// getenvoy has until tests that ensure --admin-address-path is always set. However, we can't read the
```
Comment (codefromthecrypt) [inline e2e/getenvoy_run_test.go:101]:
```suggestion
			// getenvoy has unit tests that ensure --admin-address-path is always set. However, we can't read the
```

File: e2e/getenvoy_run_test.go [hunk 5]
```diff
@@ -30,113 +33,139 @@ import (
...
+	go func() {
+		for errLines.Scan() {
+			l := errLines.Text()
+			fmt.Fprintln(os.Stderr, l) // echo to the person watching the test
 
-	// If this passes, we know Envoy is running in the current directory, so can resolve relative configuration.
-	require.Equal(t, responseFromWorkingDirectory, body, `unexpected content in %s after running [%v]`, mainURL, c)
-}
+			// When we get to this line, we can assume Envoy started properly. Run the test
+			if strings.Contains(l, "starting main dispatch loop") {
```
Comment (codefromthecrypt) [inline e2e/getenvoy_run_test.go:117]:
```suggestion
			if strings.Contains(l, envoyStartedLine) {
```

File: e2e/main_test.go [hunk 6]
```diff
@@ -124,3 +131,29 @@ func readGetEnvoyPath() (string, error) {
...
+	return &getEnvoy{exec.CommandContext(ctx, getEnvoyPath, args...), ""}, cancel //nolint:gosec
+}
+
+func (b *getEnvoy) String() string {
+	return strings.Join(b.cmd.Args, " ")
+}
+
+func getEnvoyExec(args ...string) (string, string, error) {
```
Comment (codefromthecrypt) [inline e2e/main_test.go:150]: not `getEnvoyRun` because don't want to stutter `getEnvoyRun("run" ...`

**Comment by codefromthecrypt** [comment]
Besides tech debt, some more motivation for the refactoring is windows support. The former code was hairy and difficult to troubleshoot what was happening

File: e2e/getenvoy_run_test.go [hunk 7]
```diff
@@ -30,113 +33,149 @@ import (
...
+	return c, filepath.Dir(adminAddressPath) + ".tar.gz"
+}
+
+func runTestAndInterruptEnvoy(t *testing.T, c *getEnvoy, adminAddressPath string, test func(*getEnvoy, *adminClient)) {
+	defer func() {
+		log.Printf(`shutting down Envoy after running [%v]`, c)
+		_ = c.cmd.Process.Signal(syscall.SIGTERM)
+	}()
+	a := requireEnvoyReady(t, adminAddressPath, c)
+	if test != nil {
```
Comment (mathetake) [inline e2e/getenvoy_run_test.go:145]: when does `test` become nil? is that allowed here?

**Review by mathetake — APPROVED** [review]
GJ!

File: e2e/getenvoy_run_test.go [hunk 8]
```diff
@@ -30,113 +33,149 @@ import (
...
-
-	runDir := requireRunDir(t, stdout, c)
-	requireEnvoyReady(t, runDir, stderr, c)
+	// Below is the minimal config needed to run Envoy 1.18+, non-windows <1.18 need access_log_path: '/dev/stdout'
+	args := []string{"run", "--config-yaml", "admin: {address: {socket_address: {address: '127.0.0.1', port_value: 0}}}"}
 
-	log.Printf(`shutting down Envoy after running [%v]`, c)
-	shutdown()
-	deferredShutdown = nil
+	c, runArchive := envoyRunTest(t, nil, args...)
```
Comment (codefromthecrypt) [inline e2e/getenvoy_run_test.go:53]: @mathetake sorry this is where nil is.. basically we aren't testing anything about the process before it is shutdown

**Comment by codefromthecrypt** [comment]
Thanks for night watch @mathetake!

**Pull request merged**